relays/substrate/src/headers_target.rs

```diff
@@ -90,10 +90,15 @@ where
 		let data = Bytes(Vec::new());
 
 		let encoded_response = self.client.state_call(call, data, None).await?;
-		let decoded_response: (P::Number, P::Hash) =
+		let decoded_response: Vec<(P::Number, P::Hash)> =
 			Decode::decode(&mut &encoded_response.0[..]).map_err(SubstrateError::ResponseParseFailed)?;
 
-		let best_header_id = HeaderId(decoded_response.0, decoded_response.1);
+		let best_header = decoded_response.last().ok_or_else(|| {
```

Comment:
Do we assume that the headers are sorted in some way? If we do, this assumption is not really properly documented. If we don't care about the ordering why do we waste time to encode all of them and just take the last one? The `RuntimeApi` should be tailored for our relayers use case.

Reply:
> Do we assume that the headers are sorted in some way?

Not really. The `last()` header is just the last header that's been imported into the pallet (at the best height).

> If we don't care about the ordering why do we waste time to encode all of them and just take the last one

There's [this discussion](https://github.com/paritytech/parity-bridges-common/pull/409#discussion_r509053999) where Slava goes over why the ordering doesn't matter.

> The RuntimeApi should be tailored for our relayers use case.

I agree to a certain extent, but I don't see the downside to providing a bit of flexibility with the pallet in that it can return a few of the best headers. The goal of this PR wasn't to change relayer logic, so that change was really the bare minimum to make sure we didn't break the relayer. Going forward the relayer can be updated to use multiple of the best headers in its sync procedure.

Reply:
> I don't see the downside to providing a bit of flexibility with the pallet in that it can return a few of the best headers. 

IMHO the increased maintenance cost* of the code and the fact that we've spent 3 rounds of comments discussing the purpose of this change is enough to justify that if the change is not strictly required for this PR and is there only for some quite vague "future improvements", we should not introduce it in the first place.
I.e. in the current context it introduces more confusion than adds value, especially given that pruning is not implemented and it might simply be reporting some weird data.
If we want to update the relayer to support this new sync procedure, we should just create an issue for this and then apply this change when it's really going to be used.

*It seems that currently we have some extra logic to update `BestHeaders` in `write_header`, a bunch of iterators to return in `RuntimeApi` and extra changes in the relayer (different runtime api to call), just to discard all, expect for one random header in that collection.

That said, I'm fine to keep it, just for the sake of getting the PR merged faster, but going forward I think we should really focus only on things that are actually useful at present or we have a good enough case that it is going to be needed in the near future either by us or external developers (pallet as a library).

Reply:
I still think the pallet should provide all the best headers and it's the relays choice whether or not they want to use them. Handing back one kinda arbitrary header at the pallet level doesn't seem like intuitive behaviour for anybody building on top of us. If we are going to return a sole header we need to really think about what guarantees we want to provide to users of the pallet.

I'm interested to hear @svyatonik's opinion on this since he's driving the relay development.